calculator/main.py

```diff
@@ -26,7 +43,24 @@ def perform_operation(calc, operation, params):
     :param params: Tuple containing the list of nums to operate with.
                    ie: (1, 2, 3, 4.5, -2)
     """
-    pass
+    if len(params) == 0:
+        raise InvalidParams("Parameters should not be empty")
+    for item in params:
+        if not (isinstance(item, int) or isinstance(item, float)):
+            raise InvalidParams("Given params are invalid.")
+    if operation not in calc['operations'].keys():
+        raise InvalidOperation("Given operation is invalid.")
+    
+    result = calc['operations'][operation](*params)
+    
+    dt = datetime.now()
+    date = dt.strftime("%Y-%m-%d %X")
+    history = [date, operation, params, result]
+    
+    #('2016-05-20 12:00:00', 'add', (1, 2), 3),
+    calc["history"].append(tuple(history))
```

Comment:
Instead of casting history as a tuple here you could just make it a tuple in the first place. ie.: `history = (date, operation, params, result)`